Port to VisualStudio compiler and Win-10 (not complete)

mrkkrj opened the pull request:
- added support for Microsoft compiler 
- fixed crash on Win-64
- added support for time measurements to run.py on Win-10

OPEN TODO: run.py build support for Windows!!!
 --> But if you build manually, you can use the run.py script to compare results

mrkkrj committed:
- added support for Microsoft compiler
 - fixed crash on Win-64
 - added support for time measurements to run.py on Win-10

OPEN TODO: run.py build support for Windows!!!

mrkkrj committed:
added support for CMake builds on Windows in the run.py script

OPEN TODO:: validation is failing bc. of different file sizes!!!

mrkkrj commented:
New commit added - building on Windows possible with the run.py script.
Problem -> validation fails bc of different file sizes!

DenisBakhvalov commented on run.py:
```diff
@@ -74,15 +74,29 @@
+        if sys.platform != 'win32':                
+          subprocess.check_call("cp " + inputFileName + " " + destDir, shell=True, stdout=FNULL, stderr=FNULL)
+        else:
+            subprocess.check_call("copy " + inputFileName + " " + destDir, shell=True, stdout=FNULL, stderr=FNULL)
```
```suggestion
          subprocess.check_call("copy " + inputFileName + " " + destDir, shell=True, stdout=FNULL, stderr=FNULL)
```

DenisBakhvalov commented on run.py:
```diff
@@ -108,39 +122,58 @@
+              scores.append(float(time))          
+        else:
+          print("  run - WIN 32")
+          output = subprocess.check_output("powershell -Command \"Measure-Command {" + runCmd + " | Out-Default}\" | findstr TotalSeconds" , shell=True)
```
Do you know if powershell is available by default?

mrkkrj replied:
Docs says so:

"Windows PowerShell comes installed by default in every Windows, starting with Windows 7 SP1 and Windows Server 2008 R2 SP1."

https://docs.microsoft.com/de-de/powershell/scripting/windows-powershell/install/installing-windows-powershell?view=powershell-7.1

DenisBakhvalov replied:
Ok, thanks!

DenisBakhvalov reviewed (commented):
Thanks, Marek! The changes look good.

> Problem -> validation fails bc of different file sizes!
That's interesting. Will you find the time to take a look at it, please?

mrkkrj commented:
I'll have a look today...
Mrk.

On Mon, Feb 15, 2021, 18:54 DenisBakhvalov <notifications@github.com> wrote:

> *@DenisBakhvalov* commented on this pull request.
>
> Thanks, Marek! The changes look good.
>
> Problem -> validation fails bc of different file sizes!
> That's interesting. Will you find the time to take a look at it, please?
>
> ------------------------------
>
> In run.py
> <https://github.com/dendibakh/perf_challenge4/pull/4#discussion_r576350478>
> :
>
> >
>
>        inputFileName = str(os.path.join(saveCwd, "221575.pgm"))
>
>        destDir = str(submissionBuildDir)
>
>        try:
>
> -        subprocess.check_call("cp " + inputFileName + " " + destDir, shell=True, stdout=FNULL, stderr=FNULL)
>
> +        if sys.platform != 'win32':
>
> +          subprocess.check_call("cp " + inputFileName + " " + destDir, shell=True, stdout=FNULL, stderr=FNULL)
>
> +        else:
>
> +            subprocess.check_call("copy " + inputFileName + " " + destDir, shell=True, stdout=FNULL, stderr=FNULL)
>
>
> ⬇️ Suggested change
>
> -            subprocess.check_call("copy " + inputFileName + " " + destDir, shell=True, stdout=FNULL, stderr=FNULL)
>
> +          subprocess.check_call("copy " + inputFileName + " " + destDir, shell=True, stdout=FNULL, stderr=FNULL)
>
>
> ------------------------------
>
> In run.py
> <https://github.com/dendibakh/perf_challenge4/pull/4#discussion_r576351415>
> :
>
> >        for x in range(0, 10):
>
> -        output = subprocess.check_output("time -p " + runCmd, shell=True)
>
> -        for row in output.split(b'\n'):
>
> -          if b'real' in row:
>
> -            real, time = row.split(b' ')
>
> -            scores.append(float(time))
>
> -
>
> +        if sys.platform != 'win32':
>
> +          output = subprocess.check_output("time -p " + runCmd, shell=True)
>
> +          for row in output.split(b'\n'):
>
> +            if b'real' in row:
>
> +              real, time = row.split(b' ')
>
> +              scores.append(float(time))
>
> +        else:
>
> +          print("  run - WIN 32")
>
> +          output = subprocess.check_output("powershell -Command \"Measure-Command {" + runCmd + " | Out-Default}\" | findstr TotalSeconds" , shell=True)
>
>
> Do you know if powershell is available by default?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dendibakh/perf_challenge4/pull/4#pullrequestreview-590672518>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABR5TYGEU7G5FU4KIK7PATS7FNWXANCNFSM4XUOBZOQ>
> .
>

mrkkrj committed:
corrected the CRLF problem on Windows

mrkkrj commented:
The problem was that fprintf uses CLRF on Windows!

DenisBakhvalov commented on canny_baseline/canny_source.c:
```diff
@@ -975,7 +975,11 @@ int write_pgm_image(char *outfilename, unsigned char *image, int rows,
    ***************************************************************************/
    if(outfilename == NULL) fp = stdout;
    else{
+#if _WIN32
```
@ibogosavljevic, please review Marek's changes in canny_source.c
Do they conflict with your changes in #5?

ibogosavljevic commented:
They dont conflict, I modified only validate.cpp.

All Marek's changes are inside #ifdef Win32, he didn't touch the common
code so I guess they are safe to land.

On Tue, 16 Feb 2021 at 18:18, DenisBakhvalov <notifications@github.com>
wrote:

> *@DenisBakhvalov* commented on this pull request.
> ------------------------------
>
> In canny_baseline/canny_source.c
> <https://github.com/dendibakh/perf_challenge4/pull/4#discussion_r576996954>
> :
>
> > @@ -975,7 +975,11 @@ int write_pgm_image(char *outfilename, unsigned char *image, int rows,
>     ***************************************************************************/
>     if(outfilename == NULL) fp = stdout;
>     else{
> +#if _WIN32
>
> @ibogosavljevic <https://github.com/ibogosavljevic>, please review
> Marek's changes in canny_source.c
> Do they conflict with your changes in #5
> <https://github.com/dendibakh/perf_challenge4/pull/5>?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dendibakh/perf_challenge4/pull/4#pullrequestreview-591450547>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACAJBL4PM6DSYIODPLDK3PDS7KSIFANCNFSM4XUOBZOQ>
> .
>

DenisBakhvalov commented:
Thanks, @ibogosavljevic .
@mrkkrj do you plan more changes in this PR?

mrkkrj commented:
No, at the moment I don't have time to play with it more. But the basic usage case is running - at least for me with Visual Studio 2019.

dendibakh committed:
Update run.py

Co-authored-by: DenisBakhvalov <61807338+DenisBakhvalov@users.noreply.github.com>
